src/NHibernate/Cache/StandardQueryCache.cs

```diff
@@ -335,9 +372,9 @@ public void Destroy()
 			ICacheAssembler[] returnTypes,
 			ISessionImplementor session,
 			IList result,
-			long ts)
+			long ts, string[] aliases)
 		{
-			var cacheable = new List<object>(result.Count + 1) { ts };
+			var cacheable = new List<object>(result.Count + 2) { ts, aliases };
```

Comment:
Maybe put it as single array item instead?

`var cacheable = new List<object>(result.Count + 1) {new[]{ ts, aliases} };`

Reply:
I think this would work.

Reply:
Overall it is better, but it is not enough for not breaking if mixed versions of Nhibernate put data in shared cache. the old ones will be broken, the newer one could support the old one data through some specific code, but would it be worth it?

I think we should consider this a possible breaking change, mandating to purge the cache and upgrade all application instances using the cache if it is a distributed cache.

Reply:
Putting the aliases at the end could be non-breaking for distributed cache scenario mixing NHibernate versions (or not purging the cache on upgrade), but are we willing to really supports this?

Putting them at the end in a non-breaking way would be a bit cumbersome: we would have to add them only for auto-discovered types queries, which are not cacheable with older NHibernate version, so which would not break them. But then the logic to iterate the result list would have to be different according to whether the query is with auto-discovery or not.

Reply:
>  only for auto-discovered types queries

Only for auto-discovered types queries with not null result transformer. 

Anyway we just need to add note about possible breaking change.
As I don't think it's possible to do completely in non-breaking way. It's not cacheable only since 5.3.14 (#3172). It throws in 5.3.11 - 5.3.13 on attempt to retrieve from cache (caused by #2943 which fixed cached auto discovery queries with null transformer)
But in 5.3.0 - 5.3.10 - it's cached successfully but not retrieved from cache. So mixing with  5.3.0 - 5.3.10 would fail.

I'm suggesting single reserved item just because I think it's more clean solution and it would be easier to extend it further in future.

We can also keep backward compatibility for not affected queries using something like:
`var cacheable = new List<object>(result.Count + 1) {aliases == null ? (object)ts : new[]{ ts, aliases} };`
Though I don't think we should really do it (but not against it too)..

Reply:
> Maybe we should add some test using NHibernate Caches instead

Good idea

Reply:
I think we should do what @bahusoid proposes. For the caches we’re trying to fix nothing will change (the we’re not working anyway), but for other types we would need to flush cache. 

With the changes proposed by Roman we can skip flushing.